File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    bool sameTree(TreeNode* p, TreeNode* q) {
        if (p == nullptr || q == nullptr) {
            return p == nullptr && q == nullptr;
        }
        if (p->val != q->val) {
            return false;
        }
        return sameTree(p->left, q->left) && sameTree(p->right, q->right);
    }

public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (subRoot == nullptr) {
            return true;
        }
        if (root == nullptr) {
            return false;
        }
        if (root->val == subRoot->val && sameTree(root->left, subRoot->left) && sameTree(root->right, subRoot->right)) {
            return true;
        }
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
    }
};

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def registerLevel(self, levels, currLevel, root):
        if not root:
            return
        if len(levels) == currLevel:
            levels.append([])
        levels[currLevel].append(root.val)
        self.registerLevel(levels, currLevel + 1, root.left)
        self.registerLevel(levels, currLevel + 1, root.right)
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        levels = []
        self.registerLevel(levels, 0, root)
        return levels
